### PR TITLE
Issue #4923 - restore caching of SSLSession information for SSL Attributes

### LINKED ISSUE #4923: SecureRequestCustomizer.SslAttributes does not cache cert chain like before
**Jetty version**
9.4.29

**Description**
When a long lived connection has multiple requests and something accesses the `HttpServletRequest.getAttributes("javax.servlet.request.X509Certificate")` then the entire Certificate Chain is resolved with each access of the Attributes.

The change in PR #4816 apparently got rid of the SslSession `CachedInfo` object that made this lookup efficient.

Old code in 9.4.28 looks like this...

https://github.com/eclipse/jetty.project/blob/jetty-9.4.28.v20200408/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java#L299

https://github.com/eclipse/jetty.project/blob/ab228fde9e55e9164c738d7fa121f8ac5acd51c9/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java#L289-L307

### PR DESCRIPTION
**Pull request by joakime**
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 1]
```diff
@@ -377,4 +397,41 @@ public Object getAttribute(String name)
             return names;
         }
     }
+
+    /**
+     * Simple bundle of information that is cached in the SSLSession. Stores the
+     * effective keySize and the client certificate chain.
+     */
+    private static class CachedSslSessionInfo
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Please rename the class to `SslSessionData`, and everywhere "info" to "data". Drop `Cached`.
Also make everything `private`: fields, methods, etc.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 2]
```diff
@@ -377,4 +397,41 @@ public Object getAttribute(String name)
...
+    {
+        /**
+         * The name of the SSLSession attribute that will contain any cached information.
+         */
+        public static final String ATTR = CachedSslSessionInfo.class.getName();
+        private final X509Certificate[] _certs;
+        private final Integer _keySize;
+        private final String _idStr;
+
+        CachedSslSessionInfo(Integer keySize, X509Certificate[] certs, String idStr)
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Why `Integer`? Should be `int`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 3]
```diff
@@ -377,4 +397,41 @@ public Object getAttribute(String name)
...
+     */
+    private static class CachedSslSessionInfo
+    {
+        /**
+         * The name of the SSLSession attribute that will contain any cached information.
+         */
+        public static final String ATTR = CachedSslSessionInfo.class.getName();
+        private final X509Certificate[] _certs;
+        private final Integer _keySize;
+        private final String _idStr;
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:412]: Keep field declarations in the same order as the constructor parameters and constructor assignments.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 2 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Came from the old (9.4.28) CachedInfo.

https://github.com/eclipse/jetty.project/blob/ab228fde9e55e9164c738d7fa121f8ac5acd51c9/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java#L360-L387

Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Fixed

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 1 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Fixed.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 3 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:412]: Done.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 4]
```diff
@@ -377,4 +400,40 @@ public Object getAttribute(String name)
...
+    /**
+     * Simple bundle of information that is cached in the SSLSession.
+     */
+    private static class SslSessionData
+    {
+        /**
+         * The name of the SSLSession attribute that will contain this information.
+         */
+        private static final String ATTR = SslSessionData.class.getName();
+        private final Integer _keySize;
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:410]: `int`, not `Integer`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 5]
```diff
@@ -377,4 +400,40 @@ public Object getAttribute(String name)
...
+        private final String _idStr;
+
+        private SslSessionData(int keySize, X509Certificate[] certs, String idStr)
+        {
+            this._keySize = keySize;
+            this._certs = certs;
+            this._idStr = idStr;
+        }
+
+        private X509Certificate[] getCerts()
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Put this getter after `getKeySize()` to be consistent with ordering.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 6]
```diff
@@ -377,4 +400,40 @@ public Object getAttribute(String name)
...
         }
     }
+
+    /**
+     * Simple bundle of information that is cached in the SSLSession.
+     */
+    private static class SslSessionData
+    {
+        /**
+         * The name of the SSLSession attribute that will contain this information.
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: `information` -> `data`

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 7]
```diff
@@ -377,4 +400,40 @@ public Object getAttribute(String name)
             return names;
         }
     }
+
+    /**
+     * Simple bundle of information that is cached in the SSLSession.
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: `information` -> `data`

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 8]
```diff
@@ -363,6 +362,30 @@ public Object getAttribute(String name)
...
+
+                X509Certificate[] certs = getCertChain(_request.getHttpChannel().getConnector(), _session);
+
+                byte[] bytes = _session.getId();
+                String idStr = TypeUtil.toHexString(bytes);
+
+                sslSessionData = new SslSessionData(keySize, certs, idStr);
+                _session.putValue(SslSessionData.ATTR, sslSessionData);
+            }
+            return sslSessionData;
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:387]: This method is the only user of `SslSessionData.ATTR`.
So perhaps we don't need the constant in another class.
We can do at beginning of method:

`String key = SslSessionData.class.getName()`

and then use `key` where now we use `ATTR`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 4 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:410]: Isn't `Integer` better here so it doesn't get boxed on every get?

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java [hunk 9]
```diff
@@ -377,4 +400,40 @@ public Object getAttribute(String name)
...
+        }
+
+        private X509Certificate[] getCerts()
+        {
+            return _certs;
+        }
+
+        private int getKeySize()
+        {
+            return _keySize;
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: If the `_keySize` is an `Integer` then return that here so it doesn't get unboxed and reboxed.
Use `int` everywhere or `Integer` everywhere.

**Review by gregw — CHANGES_REQUESTED** [review]
Almost LGTM.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 4 above)
Comment (sbordet) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:410]: Ah yes, `Integer` everywhere - I missed the boxing when returning the attribute, and the `keySize` values are not cached in `Integer` (only caches `-128` to`127`).

**Review by lorban — CHANGES_REQUESTED** [review]
LGTM, after the requested changes.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 9 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Done.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 8 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:387]: Done.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 7 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Done.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 6 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: This ATTR was removed.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 5 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java]: Moved.

File: jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java (hunk 4 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:410]: It's Integer everywhere now.

**Review by gregw — APPROVED** [review]

**Review by lorban — APPROVED** [review]

**Pull request merged**